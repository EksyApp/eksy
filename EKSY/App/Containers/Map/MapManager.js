import React from 'react'
import * as Actions from '../../Actions'
import Store from '../../Store'
import BackgroundGeolocation from 'react-native-mauron85-background-geolocation'

let instance = null
let idCounter = 0

class MapManager {

	constructor() {
		if (!instance) {
			this._map = null
			this._currentLocationMoveRequested = false
			this._reduxState = null
			this.initStore()
			// this.store.subscribe(() => this._storeListener())
			this.startLocationWatcher()
			instance = this
		}
		return instance
	}

	async initStore() {
		this.store = await Store()
		this.store.subscribe(() => this.storeListener())
	}

	static getNextID() {
		idCounter++
		return idCounter
	}

	storeListener() {
		this._reduxState = this.store.getState()
		this.handleFlyingToCurrentLocation()
	}

	handleFlyingToCurrentLocation() {
		if (this._currentLocationMoveRequested) {
			this.goToCurrentPosition()
		}
	}

	startLocationWatcher() {
		BackgroundGeolocation.configure({
			desiredAccuracy: 10,
      stationaryRadius: 10,
      distanceFilter: 10,
      locationTimeout: 30,
      debug: false,
      startOnBoot: false,
      stopOnTerminate: true,
      locationProvider: BackgroundGeolocation.provider.ANDROID_ACTIVITY_PROVIDER,
      interval: 10000,
      fastestInterval: 5000,
      activitiesInterval: 10000,
      stopOnStillActivity: true
		})

		BackgroundGeolocation.on('location', (location) => {
				this.store.dispatch(Actions.updateLocation(location))
				this.store.dispatch(Actions.locationKnown(true))
				this._map.forceUpdate()
		})

		BackgroundGeolocation.on('stationary', () => {})

		BackgroundGeolocation.on('error', (error) => {
			console.log('Geolocation error: ' + error)
			this.store.dispatch(Actions.locationKnown(false))
		})

		BackgroundGeolocation.start(() => {
			console.log('BackgroundGeolocation started')
		})
		// this.watchID = navigator.geolocation.watchPosition(
		// 		(position) => {
		// 			this.store.dispatch(Actions.updateLocation(position.coords))
		// 			this.store.dispatch(Actions.locationKnown(true))
		// 			this._map.forceUpdate()
		// 		},
		// 		(error) => this.store.dispatch(Actions.locationKnown(false)),
		// 		{enableHighAccuracy: false, timeout: 500, maximumAge: 0, distanceFilter: 3}
		// )
	}

	setMapObject(map) {
		this._map = map
	}

	goToCurrentPosition() {
		if (this._reduxState && this._reduxState.map.location.isKnown) {
			this._currentLocationMoveRequested = false
			this.flyToPosition(this._reduxState.map.location.latitude, this._reduxState.map.location.longitude)
		} else {
			this._currentLocationMoveRequested = true
		}
	}

	flyToPosition(latitude, longitude) {
		let position = {
			latitude: latitude,
			longitude: longitude
		}
		this._map.animateToCoordinate(position, 1000)
	}

}

export default MapManager
