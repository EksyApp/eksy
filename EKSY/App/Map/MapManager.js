import React from 'react'
import * as Actions from '../Actions'
import configureStore from '../Store'
import Marker from './Marker'

let instance = null
let idCounter = 0

class MapManager {

  constructor () {
    if (!instance) {
      this._markers = new Map()
      this._map = null
      this._currentLocationMoveRequested = false
      this._reduxState = null
      this.initStore()
      // this.store.subscribe(() => this.storeListener())
      this.startLocationWatcher()
      instance = this
    }
    return instance
  }

  async initStore () {
    this.store = await configureStore()
    this.store.subscribe(() => this.storeListener())
  }

  static getNextID () {
    idCounter++
    return idCounter
  }

  storeListener () {
    this._reduxState = this.store.getState()
    this.handleFlyingToCurrentLocation()
  }

  handleFlyingToCurrentLocation () {
    if (this._currentLocationMoveRequested) {
      this.goToCurrentPosition()
    }
  }

  startLocationWatcher () {
    navigator.geolocation.watchPosition(
				(position) => {
  this.store.dispatch(Actions.updateLocation(position.coords))
  this.store.dispatch(Actions.locationKnown(true))
  this._map.forceUpdate()
},
				(error) => this.store.dispatch(Actions.locationKnown(false)),
				{enableHighAccuracy: false, timeout: 500, maximumAge: 500, distanceFilter: 10}
		)
  }

  addMarker (key, marker) {
    let markerComponent = <Marker title={marker.title} images={marker.images} latitude={marker.latitude} longitude={marker.longitude} text={marker.text} key={MapManager.getNextID()} />
    this._markers.set(key, markerComponent)
    this._map.forceUpdate()
  }

  removeMarker (key) {
    this._markers.delete(key)
    this._map.forceUpdate()
  }

  getMarkers () {
    return this._markers
  }

  setMapObject (map) {
    this._map = map
  }

  goToCurrentPosition () {
    if (this._reduxState && this._reduxState.map.location.isKnown) {
      this._currentLocationMoveRequested = false
      this.flyToPosition(this._reduxState.map.location.latitude, this._reduxState.map.location.longitude)
    } else {
      this._currentLocationMoveRequested = true
    }
  }

  flyToPosition (latitude, longitude) {
    let position = {
      latitude: latitude,
      longitude: longitude
    }
    this._map.animateToCoordinate(position, 1000)
  }

}

export default MapManager
