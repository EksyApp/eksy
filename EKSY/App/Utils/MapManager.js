import React from 'react'
import * as Actions from '../Actions/index'
import Store from '../Store/index'
import Dao from "../Dao/Dao";
import {PermissionsAndroid} from 'react-native'
//import BackgroundGeolocation from 'react-native-mauron85-background-geolocation'

let instance = null

// used to give commands to the Google Map
class MapManager {

	constructor() {
		if (!instance) {
			this._map = null
			this._reduxState = null
			this.dao = new Dao()
			this.initStore()
			instance = this
		}
		return instance
	}

	async initStore() {
		this.store = await Store()
		this.store.subscribe(() => this.storeListener())
		this.startLocationWatcher()
	}

	storeListener() {
		this._reduxState = this.store.getState()
	}

	async startLocationWatcher() {
		this._reduxState = this.store.getState()
		
		if(this._reduxState.map.location.isKnown && this._reduxState.map.location.latitude != null) {
			this.dao.updateLocation(this._reduxState.map.location.latitude, this._reduxState.map.location.longitude)
		}
		
		try {
			const granted = await PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION)
			if (granted === PermissionsAndroid.RESULTS.GRANTED) {
				this.watchID = navigator.geolocation.watchPosition(
						(position) => {
							this.dao.updateLocation(this._reduxState.map.location.latitude, this._reduxState.map.location.longitude)
							this.store.dispatch(Actions.updateLocation(position.coords))
							this.store.dispatch(Actions.locationKnown(true))
							this._map.forceUpdate()
						},
						(error) => this.store.dispatch(Actions.locationKnown(false)),
						{enableHighAccuracy: false, timeout: 500, maximumAge: 0, distanceFilter: 3}
				)
			} else {
				this.store.dispatch(Actions.locationKnown(false))
			}
		} catch (err) {
			this.store.dispatch(Actions.locationKnown(false))
		}
		
		
	}

	setMapObject(map) {
		this._map = map
	}

	goToCurrentPosition() {
		this.flyToPosition(this._reduxState.map.location.latitude, this._reduxState.map.location.longitude)
	}

	flyToPosition(latitude, longitude) {
		let position = {
			latitude: latitude,
			longitude: longitude
		}
		this._map.animateToCoordinate(position, 1000)
	}

}

export default MapManager
