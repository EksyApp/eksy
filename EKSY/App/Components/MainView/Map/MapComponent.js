import React, {Component} from 'react'
import {
	StyleSheet,
	View,
	Animated,
	Dimensions
} from 'react-native'
import MapView from 'react-native-maps'
import MapManager from '../../../Utils/MapManager'
import Marker from './Marker'
import PropTypes from 'prop-types'
import {LocationShape, MarkerShape, MarkersShape, RegionShape, RouteShape} from '../../../Utils/PropTypeShapes'
import GeoFire from 'geofire'

const Screen = {
	width: Dimensions.get('window').width,
	height: Dimensions.get('window').height - 75
}

//Renders map view inside the main view (MainViewComponent)
//Map component holds clickable markers (Marker)
export default class MapComponent extends Component {
	
	constructor(props) {
		super(props)
		this._manager = new MapManager()
		this._manager.setMapObject(this)
		this._map = null
		
		this.state = {
			zoomLevel: (360 * ((Screen.width / 256) / this.props.currentRegion.longitudeDelta)) + 1
		}
	}
	
	animateToCoordinate(position, delay) {
		if (this._map) {
			this._map._component.animateToCoordinate(position, delay)
		}
	}
	
	handleRegionChange(region) {
		this.props.regionChange(region)
		this.setState({zoomLevel: (360 * ((Screen.width / 256) / region.longitudeDelta)) + 1})
	}
	
	renderMarkers() {
		return this.props.markerList.map((marker, index) =>
				<Marker
						data={marker}
						setMarkerSelected={this.props.setMarkerSelected}
						setMarkerViewVisible={this.props.setMarkerViewVisible}
						disableGestures={this.props.disableGestures}
						key={marker.key}
				/>
		)
		
	}
	
	
	renderUserCircle() {
		if (this.props.currentLocation.isKnown) {
			return (
					<Animated.View>
						<MapView.Circle center={this.props.currentLocation}
						                radius={this.props.radius * 1000}
						                strokeWidth={0.5}
						                strokeColor="rgba(66, 180, 230, 1)"
						                fillColor="rgba(66, 180, 230, 0.2)"
						/>
						<MapView.Circle center={this.props.currentLocation}
						                radius={300000 / this.state.zoomLevel}
						                strokeWidth={0.5}
						                strokeColor="rgba(66, 180, 230, 1)"
						                fillColor="rgba(66, 180, 230, 1)"
						/>
					</Animated.View>
			)
		}
		
		return null
	}
	
	renderRouteLine() {
		if (this.props.routeIsActive) {
			let distance = GeoFire.distance([this.props.currentLocation.latitude, this.props.currentLocation.longitude], [this.props.nextMarker.latitude, this.props.nextMarker.longitude])
			console.log(distance)
			console.log(this.props.radius)
			if (distance >= this.props.radius) {
				let latitude = this.props.currentLocation.latitude + (this.props.radius / distance) * (this.props.nextMarker.latitude - this.props.currentLocation.latitude)
				let longitude = this.props.currentLocation.longitude + (this.props.radius / distance) * (this.props.nextMarker.longitude - this.props.currentLocation.longitude)
				return (<MapView.Polyline coordinates={[this.props.currentLocation, {latitude, longitude}]}/>)
			} else {
				return (<MapView.Polyline coordinates={[this.props.currentLocation, this.props.nextMarker]}/>)
			}
		}
	}
	
	renderMapView() {
		return (
				<MapView.Animated
						ref={(ref) => this._map = ref}
						style={styles.map}
						initialRegion={this.props.currentRegion}
						onRegionChange={(region) => this.handleRegionChange(region)}
						followUserLocation
						showsMyLocationButton={true}
						showsBuildings={true}
						showCompass={false}
				>
					{this.renderRouteLine()}
					{this.renderUserCircle()}
					{this.renderMarkers()}
				</MapView.Animated>
		)
	}
	
	render() {
		return (
				<View style={styles.container}>
					{this.renderMapView()}
				</View>
		)
	}
	
	
}


const styles = StyleSheet.create({
	container: {
		flex: 1,
		justifyContent: 'flex-end',
		alignItems: 'center',
	},
	map: {
		position: 'absolute',
		top: 0,
		bottom: 0,
		left: 0,
		right: 0,
	},
	buttonContainer: {
		marginVertical: 20,
	},
	button: {
		alignItems: 'center',
		backgroundColor: 'rgba(255,255,255,0.7)',
		borderRadius: 20,
		padding: 12,
		width: 160,
	},
	callout: {
		width: 140,
	},
	circle: {
		borderRadius: 12,
		backgroundColor: 'rgba(130,4,150, 0.3)',
		position: 'absolute',
		borderWidth: 1,
		borderColor: 'rgba(130,4,150, 0.5)'
	}
})

MapComponent.propTypes = {
	currentRegion: RegionShape,
	currentLocation: LocationShape,
	radius: PropTypes.number,
	markerList: MarkersShape,
	setMarkerSelected: PropTypes.func,
	setMarkerViewVisible: PropTypes.func,
	disableGestures: PropTypes.func,
	routeIsActive: PropTypes.bool,
	route: RouteShape,
	nextMarker: MarkerShape
}

