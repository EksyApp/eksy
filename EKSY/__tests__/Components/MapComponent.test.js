import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import MapComponent from "../../App/Components/MainView/Map/MapComponent";
import MapManager from "../../App/Utils/MapManager";

jest.mock('react-native-fetch-blob', () => {
	return {
		DocumentDir: () => {},
		polyfill: () => {}
	}
})

MapManager.prototype.startLocationWatcher = jest.fn()
MapManager.prototype.storeListener = jest.fn()
MapManager.prototype.goToCurrentPosition = jest.fn()
MapManager.prototype.flyToPosition = jest.fn()

jest.mock('react-native-maps', () => {
	const React = require.requireActual('react');
	const MapView = require.requireActual('react-native-maps');
	
	class MockCallout extends React.Component {
		render() {
			return React.createElement('Callout', this.props, this.props.children);
		}
	}
	
	class MockMarker extends React.Component {
		render() {
			return React.createElement('Marker', this.props, this.props.children);
		}
	}
	
	class MockMapView extends React.Component {
		render() {
			return React.createElement('MapView', this.props, this.props.children);
		}
	}
	
	class MockMapViewAnimated extends React.Component {
		render() {
			return React.createElement('MapView', this.props, this.props.children);
		}
	}
	
	MockCallout.propTypes = MapView.Callout.propTypes;
	MockMarker.propTypes = MapView.Marker.propTypes;
	MockMapView.propTypes = MapView.propTypes;
	MockMapView.Marker = MockMarker;
	MockMapView.Callout = MockCallout;
	MockMapView.Animated = MockMapViewAnimated;
	return MockMapView;
});

test('MapComponent renders correctly', () => {
	const tree = renderer.create(
			<MapComponent
					currentRegion={{
						latitude: 1,
						longitude: 1,
						latitudeDelta: 1
					}}
			    
			    currentLocation={{
			    	isKnown: false
			    }}
			    
			    markerList={[]}
			/>
	).toJSON();
	expect(tree).toMatchSnapshot();
});