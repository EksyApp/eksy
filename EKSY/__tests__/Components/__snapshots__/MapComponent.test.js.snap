// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapComponent renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "flex-end",
    }
  }
>
  <MapView
    followUserLocation={true}
    initialRegion={
      Object {
        "latitude": 1,
        "latitudeDelta": 1,
        "longitude": 1,
      }
    }
    onRegionChange={[Function]}
    showCompass={false}
    showsBuildings={true}
    showsMyLocationButton={true}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
